File: 1140-stone-game-ii/1140-stone-game-ii.py
class Solution:
    def stoneGameII(self, piles):
        n = len(piles)

        # suffix[i] = sum of piles from i to end
        suffix = [0] * (n + 1)

        for i in range(n - 1, -1, -1):
            suffix[i] = suffix[i + 1] + piles[i]

        # dp[i][M]
        dp = [[-1] * (n + 1) for _ in range(n)]

        def solve(i, M):

            # No piles left
            if i == n:
                return 0

            # Already calculated
            if dp[i][M] != -1:
                return dp[i][M]

            best = 0

            # Try taking X piles
            for X in range(1, 2 * M + 1):

                if i + X > n:
                    break

                newM = max(M, X)

                # My stones = remaining - opponent's stones
                current = suffix[i] - solve(i + X, newM)

                best = max(best, current)

            dp[i][M] = best
            return best

        return solve(0, 1)

File: 1140-stone-game-ii/1140-stone-game-ii.java
class Solution {

    int n;
    int[][] dp;
    int[] suffix;

    int solve(int i, int M) {

        // No piles left
        if (i == n) {
            return 0;
        }

        // Already calculated
        if (dp[i][M] != -1) {
            return dp[i][M];
        }

        int best = 0;

        // Try taking X piles
        for (int X = 1; X <= 2 * M && i + X <= n; X++) {

            int newM = Math.max(M, X);

            // My stones = remaining - opponent's stones
            int current = suffix[i] - solve(i + X, newM);

            best = Math.max(best, current);
        }

        dp[i][M] = best;

        return best;
    }

    public int stoneGameII(int[] piles) {

        n = piles.length;

        suffix = new int[n + 1];

        // Calculate suffix sums
        for (int i = n - 1; i >= 0; i--) {
            suffix[i] = suffix[i + 1] + piles[i];
        }

        dp = new int[n][n + 1];

        // -1 means not calculated
        for (int i = 0; i < n; i++) {
            java.util.Arrays.fill(dp[i], -1);
        }

        return solve(0, 1);
    }
}